Docs restructure ICP features (#4955)
dfinity/motoko: doc/md/motoko-style-guide.md

    ```

    Rationale: Due to subtyping, inferring the type from the initialization would not necessarily deduce the intended type. For example, `balance` would have type [`Nat`](../base/Nat.md) without the annotation, ruling out assignments of integers.

- Put type annotations on all public fields in a class.

    ``` motoko no-repl
    class C(init_ : Nat) {
      public let init : Nat = init_;
      public var count : Nat = 0;
    }
    ```

- Omit return type annotations of functions when the type is `()`.

    ``` motoko no-repl
    func twiceF() { f(); f() };  // no need to write ": ()"
    ```

- Omit type annotations on functions when they are passed as arguments.

    ``` motoko no-repl
    Array.map<Nat, Nat>(func n {n + 1}, a);
    ```

- Put type annotations on definitions that involve numeric types other than [`Nat`](../old-docs/base/Nat.md) or [`Int`](../old-docs/base/Int.md), to resolve the overloading between arithmetic operators and constants.

    ``` motoko no-repl
    let mask : Nat32 = 0xfc03_ff00;
    let offset : Nat32 = size + 1;
    ```

### Picking types

- Use [`Nat`](../old-docs/base/Nat.md) for any integral value that cannot be negative.

> - Use fixed-width `NatN` or `IntN` only when storing many values and space usage matters, when bit-fiddling requires the low-level interpretation of a number as a vector of bits or when matching types imposed by external requirements, such as other canisters.

Reviewer: ```suggestion
- Use fixed-width `NatN` or `IntN` only when storing many values and space usage matters, when bit-fiddling requires the low-level interpretation of a number as a vector of bits, or when matching types imposed by external requirements such as other canisters.
```

- Avoid proliferation of option types, and therefore `null`.

    Limit their use to as small a scope as possible. Rule out the `null` case and use non-option types wherever possible.

- Consider using records instead of tuples when there are more than 2 or 3 components. Records are just simple objects with named fields.

>     Note that record types need not be declared but can be used in place.

Reviewer: ```suggestion
    Note that record types can be used in place and do not need to be declared.
```

    ``` motoko no-repl
      func nodeInfo(node : Node) : {parent : Node; left : Node; right : Node} { ... }
    ```

- Consider using variants instead of [`Bool`](../old-docs/base/Bool.md) to represent binary choices.

>     Note that variant types need not be declared but can be used in place.

Reviewer: ```suggestion
    Note that variant types can be used in place and do not need to be declared.
```

    ``` motoko no-repl
    func capitalization(word : Text) : {#upper; #lower} { ... }
    ```

- Where possible, use return type `()` for functions whose primary purpose is to mutate state or cause other side effects.

    ``` motoko no-repl
    class Set<X>() {
      public func add(x : X) { ... };
      public func remove(x : X) { ... };
      ...
    };
    ```

- Consider using a record (an object with just data) as argument for long parameter lists.

    ``` motoko no-repl
    func process({seed : Float; delta : Float; data : [Record]; config : Config}) : Thing {
      ...
    };

    process{config = Config(); data = read(); delta = 0.01; seed = 1.0};
    ```

    Rationale: This expresses named parameters. This way, arguments can be freely reordered at the call site and callers are prevented from accidentally passing them in the wrong order.

- Higher-order functions, such as functions that take a callback argument, should put the function parameter last.

>     Rationale: Makes call sites more readable, and in the absence of currying, there is no point in putting the function first, like you often would in functional languages.

Reviewer: ```suggestion
    Rationale: This makes call sites more readable, and in the absence of currying, there is no point in putting the function first, like you often would in functional languages.
```

- Do not use sentinel values, such as `-1`, to represent invalid values.

    Use the option type instead.

    ``` motoko no-repl
    func lookup(x : key) : ?Nat { ... }
    ```

- Data is immutable in Motoko unless explicitly stated otherwise.

    Use mutability types and definitions (`var`) with care and only where needed.

    Rationale: Mutable data cannot be communicated or share across actors. It is more error-prone and much more difficult to formally reason about, especially when concurrency is involved.

## Features

### Statements

- Use `for` loops instead of `while` loops for iterating over a numeric range or a container.

    ``` motoko no-repl
    for (i in Iter.range(1, 10)) { ... };
    for (x in array.values()) { ... };
    ```

    Rationale: For loops are less error-prone and easier to read.

> - Use `if` or `switch` as expressions where appropriate.

Reviewer: This is too broad; what is considered "appropriate" in this context?

    ``` motoko no-repl
    func abs(i : Int) : Int { if (i < 0) -i else i };

    let delta = switch mode { case (#up) +1; case (#dn) -1 };
    ```

> - Motoko requires that all expressions in a block have type `()`, in order to prevent accidentally dropped results.

Reviewer: ```suggestion
- Motoko requires that all expressions in a block have type `()` to prevent accidentally dropped results.
```

    Use `ignore` to explicitly drop results. Do *not* use `ignore` when it’s not needed.

    ``` motoko no-repl
    ignore async f();  // fire of a computation
    ```

> - Motoko allows to omit the `return` at the end of a function, because a block evaluates to its last expression.

Reviewer: ```suggestion
- Motoko allows omitting the `return` at the end of a function because a block evaluates to its last expression.
```

>     Use this when a function is short and in "functional" style, that is, the function does not contain complex control flow or side effects.

Reviewer: ```suggestion
    Omit `return` when a function is short and in "functional" style, that is, the function does not contain complex control flow or side effects.
```

    Use explicit `return` at the end when the function contains other `return` statements or imperative control flow.

    ``` motoko no-repl
    func add(i : Nat, j : Nat) : Nat { i + j };

    func foo(a : Float, b : Float) : Float {
      let c = a*a + b*b;
      c + 2*c*c;
    };

    func gcd(i : Nat, j : Nat) : Nat {
      if (j == 0) i else gcd(j, i % j);
    };

    func gcd2(i : Nat, j : Nat) : Nat {
      var a = i;
      var b = j;
      while (b > 0) {
        let c = a;
        a := b;
        b := c % b;
      };
      return a;
    };
    ```

### Objects and records

- Use the short-hand object syntax `{x1 = e1; …​ ; xN = eN}` when using objects as simple records, i.e., data structures with no private state and no methods.

- Use `object` when creating singleton objects.

- Limit the use of objects to records where possible.

    Rationale: Only records can be sent as message parameters or results and can be stored in stable variables. Objects with methods are also more expensive to create and represent in memory.

- Use full objects only as a means for encapsulating state or behavior.

### Classes

- Use `class` to create multiple objects of the same shape.

> - Name classes after their conceptual functionality, not their implementation, except when having to distinguish multiple different implementations of the same concept. For example, `OrderedMap` vs `HashMap`).

Reviewer: ```suggestion
- Name classes after their conceptual functionality, not their implementation, except when having to distinguish multiple different implementations of the same concept. For example, `OrderedMap` vs `HashMap`.
```

- Classes are both type definitions and factory functions for objects.

    Do not use classes unless both these roles are intended; use plain type aliases or functions returning an object in other cases.

- Do not overuse classes.

    Use a module defining a plain type and functions on it where appropriate. Use classes only as a means for encapsulating state or behavior.

    Rationale: Objects with methods have disadvantages over simple record types with separate functions (see above).

> - If values of a class are meant to be sendable (shared), the class needs to provide a pair of `share`/`unshare` methods that convert to/from a sharable representation, for example, as a record.

Reviewer: ```suggestion
- If values of a class are meant to be shared, the class needs to provide a pair of `share`/`unshare` methods that convert to/from a sharable representation, for example, as a record.
```

    :::note

    For immutable classes it may seem more natural to make `unshare` a kind of static function. However, even for immutable ones it may depend on constructor arguments (such as an ordering function), so that the a pattern like `Map(compareInt).unshare(x)` seems appropriate.

    :::

> - For the time being, avoid overloading classes with too many methods, since that is currently expensive.

Reviewer: ```suggestion
- Avoid overloading classes with too many methods, since that is expensive.
```

>     Restrict to a sufficiently small set of canonical methods and make less essential ones that can be implemented on top of those into functions in the enclosing module.

Reviewer: ```suggestion
    Restrict to a sufficiently small set of canonical methods. Make less essential methods that can be implemented on top of the canonical methods into functions in the enclosing module.
```

- Use modules for "static" classes or methods.

### Modules

- Use `module` to group definitions, including types, and create a name space for them.

- Where applicable, name modules after the main type or class they implement or provide functions for.

- Limit each module to a single main concept/type/class or closely entangled family of concepts/types/classes.

### Error handling

> - Use `Result` for error handling where applicable

Reviewer: ```suggestion
- Use `Result` for error handling where applicable.
```

  - Use `Result<T, E>` rather than exceptions for recoverable errors.
  - Use meaningful error types instead of generic `Text`.

    ```motoko no-repl
    type MyError = { #NotFound; #InvalidInput; #PermissionDenied };

    func safeDivide(a : Int, b : Int) : Result<Int, MyError> {
      if (b == 0) { return #err(#InvalidInput) };
      return #ok(a / b);
    };
    ```

> - Use `assert` only for programming errors

Reviewer: ```suggestion
- Use `assert` only for programming errors.
```

  - `assert` should be reserved for conditions that should never fail if the program is correct.

    ```motoko no-repl
    assert(x > 0);  // Only use if x should never be <= 0
    ```

### Concurrency

> - Use `await` judiciously

Reviewer: ```suggestion
- Use `await` judiciously.
```

  - Minimize the number of `await` calls in a function to avoid unnecessary suspensions.

    ```motoko no-repl
    let user = await db.getUser(id); // Good ~4s
    let orders = await db.getOrders(user); // Bad: sequential calls introduce high latency ~6s
    ```

>   - Instead, parallelize calls when possible:

Reviewer: ```suggestion
  - Instead, use parallel calls when possible:
```

    ```motoko no-repl
    let (user, orders) = await (db.getUser(id), db.getOrders(id));
    ```

> - Use composite queries where applicable.

Reviewer: This is too broad and needs more context

### Pattern matching

> - Use `switch` instead of manual conditionals for variants

Reviewer: ```suggestion
- Use `switch` instead of manual conditionals for variants.
```

  - Prefer exhaustive pattern matching to avoid missing cases.
  - Exhaust all cases before using the wildcard pattern.

    ```motoko no-repl
    switch (response) {
      case (#possiblity1) { processPossibility1() };
      case (#possiblity2) { processPossibility2() }
      case (#possiblity2) { processPossibility3() };
      case (_) { handleExecption(err) };;
    };
    ```

### Data structures

> - Prefer `Buffer` over dynamic array resizing

Reviewer: ```suggestion
- Prefer `Buffer` over dynamic array resizing.
```

  - `Array` in Motoko is immutable, so if dynamic resizing is needed, `Buffer` is a better choice.

    ```motoko no-repl
    let buf = Buffer<Nat>(10);
    buf.add(42);
    let arr = Buffer.toArray<Nat>(buf);
    ```

## Motoko grammar

> This section describes the concrete syntax, or grammar, of Motoko. The specification is auto-generated with a tool. This grammar is is in the [Extended Backus–Naur form(EBNF)](https://tomassetti.me/ebnf/).

Reviewer: ```suggestion
This section describes the concrete syntax (grammar) of Motoko. The specification is auto-generated with a tool. This grammar is in the [Extended Backus–Naur form(EBNF)](https://tomassetti.me/ebnf/).
```
